README.md
# nbtty

`nbtty` runs a command with a pty that discards output from the command rather
than blocking.  This is useful for commands that log output and for whatever
reason output is redirected to a location that can't keep up.  Rather than
slowing or stopping the program, it can just keep going none the wiser.

This program also polls the terminal for its dimensions using ANSI escape
sequences. This is needed to pass on the size over a serial console since
SIGWINCH doesn't work.

The whole point of this program to make the Elixir IEx console in Nerves
projects work well. The blocking issue became especially problematic when the
IEx console was redirected over a USB gadget serial port.

This is a substantially trimmed down and modified fork of
[dtach](https://github.com/crigler/dtach). A huge thanks goes out to Ned T.
Crigler for this useful tool. Ned - if you're reading this and don't recognize
your code, sorry, I kind of butchered it.

## Usage

```sh
nbtty [--tty <tty path>|--wait-input] <command> [args...]
```

Specify `--tty` for `nbtty` to use a specific tty instead of stdin/stdout. It
will additionally retry opening the tty if it doesn't exist on start. This is
useful for getting around the problem where Elixir code initializes a tty that
provides the main console.

Specify `--wait-input` to not send any output to the tty until a carriage return
is received (user presses the enter key). This is useful if don't expect anyone
to be at the console and want to minimize the risk of garbage being received and
processed.

## Building

You need some build tools for this project. In Ubuntu, you can install them
with the following:

```sh
sudo apt install build-essential automake autoconf
```

To build:

```sh
./autogen.sh
./configure
make
```

At the end of the commands, you will have `nbtty` in the project root directory.

## License

Since `nbtty` derives from `dtach`, much of it is Copyright © 2004-2016 Ned T.
Crigler.  Like `dtach`, `nbtty` is distributed under the General Public License.
See [COPYING](COPYING) for details.


CHANGELOG.md
# Changelog

## v0.4.1

* Bug fixes
  * Add option to not write to the terminal until the user presses enter. This
    is useful for the USB interface where sometimes characters written to the
    output get looped back.
  * Switch to more traditional automake setup so that `make dist` and `make
    install` work

## v0.4.0

* New features
  * Support `--tty` option to use a specified tty device. This will be retried
    if it doesn't exist at start. Needed for configfs configuration of the
    gadget USB interface.

* Bug fixes
  * Exit on SIGCHLD from the launched process. It appears to be unreliable to
    detect exit based on tty EOF.

## v0.3.2

* Fix missing include with Musl

## v0.3.1

* Reduce window size polling by triggering it on input from the user rather than
  output from Elixir. This reduces the bytes coming out quite a bit and the
  chance that one is dropped and the ANSI escape sequence gets corrupted.

## v0.3.0

* More code simplifications

## v0.2.0

* New features
  * Added ANSI window size polling

* Bug fixes
  * Simplified tty handling code - seems to work better

## v0.1.0

* Initial release


ansi.c
#include "nbtty.h"
#include "ansi.h"

#include <string.h>

/*
 * This escape sequence does the following:
 * 1. Saves the cursor position
 * 2. Put the cursor at 999,999
 * 3. Get where the cursor ended up
 * 4. Put the cursor back
 *
 * The remote terminal should send back:
 * ESC"[%u;%uR" with the number of rows and columns.
 */
#define ESC "\033"
static const char window_size_sequence[] = ESC"7" ESC"[r" ESC"[999;999H" ESC"[6n" ESC"8";

struct ansi_parser {
    unsigned char buffer[64];
    int index;
    int state;

    int row;
    int col;

    struct winsize ws;
};
static struct ansi_parser ansi_parser;

struct ansi_parse_transition {
    const char *chars;
    int next_state;
    void (*handler)(unsigned char in, unsigned char **out);
};

struct ansi_parse_state {
    struct ansi_parse_transition transitions[3];
};

static void ansi_handle_start(unsigned char in, unsigned char **out)
{
    (void) out;
    ansi_parser.row = 0;
    ansi_parser.col = 0;
    ansi_parser.buffer[0] = in;
    ansi_parser.index = 1;
}

static void ansi_handle_capture(unsigned char in, unsigned char **out)
{
    (void) out;
    ansi_parser.buffer[ansi_parser.index++] = in;
}

static void ansi_handle_row(unsigned char in, unsigned char **out)
{
    ansi_parser.row = ansi_parser.row * 10 + ansi_parser.buffer[ansi_parser.index] - '0';
    ansi_handle_capture(in, out);
}

static void ansi_handle_col(unsigned char in, unsigned char **out)
{
    ansi_parser.col = ansi_parser.col * 10 + ansi_parser.buffer[ansi_parser.index] - '0';
    ansi_handle_capture(in, out);
}

static void ansi_handle_rc(unsigned char in, unsigned char **out)
{
    (void) in;
    (void) out;

    // set to row, col
    ansi_parser.ws.ws_row = ansi_parser.row;
    ansi_parser.ws.ws_col = ansi_parser.col;
    ansi_parser.ws.ws_xpixel = 0;
    ansi_parser.ws.ws_ypixel = 0;

    // Discard ANSI code
    ansi_parser.index = 0;
}

static void ansi_handle_mismatch(unsigned char in, unsigned char **out)
{
    if (ansi_parser.index > 0) {
        memcpy(*out, ansi_parser.buffer, ansi_parser.index);
        *out += ansi_parser.index;
        ansi_parser.index = 0;
    }
    **out = in;
    *out += 1;
}

/*
 * This is a little DFA for parsing the one ANSI response that
 * we're interested in. It is also discarded so that it doesn't
 * mess up Erlang's ANSI parser which doesn't seem to handle it.
 *
 * Note that this DFA buffers characters due to the requirement
 * to toss the ANSI response of interest. Since buffering input
 * could really confuse a user, it passes it on as soon as there
 * is no chance of a match.
 */
#define CATCH_ALL {NULL, 0, ansi_handle_mismatch}
static struct ansi_parse_state ansi_states[] = {
    /* 0 - Start */ {{{ESC, 1, ansi_handle_start}, CATCH_ALL}},
    /* 1 - Esc   */ {{{"[", 2, ansi_handle_capture}, CATCH_ALL}},
    /* 2 - [     */ {{{"0123456789", 3, ansi_handle_row}, CATCH_ALL}},
    /* 3 - n1    */ {{{"0123456789", 4, ansi_handle_row}, {";", 6, ansi_handle_capture}, CATCH_ALL}},
    /* 4 - n2    */ {{{"0123456789", 5, ansi_handle_row}, {";", 6, ansi_handle_capture}, CATCH_ALL}},
    /* 5 - n3    */ {{{";", 6, ansi_handle_capture}, CATCH_ALL}},
    /* 6 - ;     */ {{{"0123456789", 7, ansi_handle_col}, CATCH_ALL}},
    /* 7 - m1    */ {{{"0123456789", 8, ansi_handle_col}, {"R", 0, ansi_handle_rc}, CATCH_ALL}},
    /* 8 - m2    */ {{{"0123456789", 9, ansi_handle_col}, {"R", 0, ansi_handle_rc}, CATCH_ALL}},
    /* 9 - m3    */ {{{"R", 0, ansi_handle_rc}, CATCH_ALL}},
};

/**
 * Scan the input looking for window size updates. This function
 * caches input temporarily if it looks like a relevent escape
 * sequence, so the amount that should be output won't always be
 * the same as what comes in.
 *
 * If the window size changes, this will return something non-zero.
 */
int ansi_process_input(const unsigned char *input, size_t input_size,
                       unsigned char *output, size_t *output_size,
                       struct winsize *ws)
{
    unsigned char *out = output;
    for (size_t i = 0; i < input_size; i++) {
        struct ansi_parse_transition *transition = ansi_states[ansi_parser.state].transitions;
        unsigned char in = input[i];
        ansi_parser.buffer[ansi_parser.index] = input[i];
        for (;;) {
            if (transition->chars == NULL || strchr(transition->chars, input[i])) {
                // match
                transition->handler(in, &out);
                ansi_parser.state = transition->next_state;
                break;
            } else {
                // no match -> try the next one.
                transition++;
            }
        }
    }
    *output_size = (size_t) (out - output);

    if (ansi_parser.ws.ws_col != 0 &&
            ansi_parser.ws.ws_col != ws->ws_col &&
            ansi_parser.ws.ws_row != ws->ws_row) {
        *ws = ansi_parser.ws;
        return 1;
    } else {
        return 0;
    }
}

void ansi_reset_parser()
{
    memset(&ansi_parser, 0, sizeof(ansi_parser));
}

/**
 * Copy the ANSI sequence to get the window size into a buffer
 */
size_t ansi_size_request(unsigned char *dest)
{
    memcpy(dest, window_size_sequence, sizeof(window_size_sequence) - 1);
    return sizeof(window_size_sequence) - 1;
}


dtach.h
/*
    dtach - A simple program that emulates the detach feature of screen.
    Copyright (C) 2001, 2004-2016 Ned T. Crigler

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#ifndef dtach_h
#define dtach_h

#include <config.h>

#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_PTY_H
#include <pty.h>
#endif

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifdef HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif

#include <termios.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <err.h>

extern struct termios orig_term;

/*
** The master sends a simple stream of text to the attaching clients, without
** any protocol. This might change back to the packet based protocol in the
** future. In the meantime, however, we minimize the amount of data sent back
** and forth between the client and the master. BUFSIZE is the size of the
** buffer used for the text stream.
*/
#define BUFSIZE 4096

/* This hopefully moves to the bottom of the screen */
#define EOS "\033[999H"

int attach_main(int s);
int master_main(char **argv, int s);

#endif


nbtty.h
/*
    dtach - A simple program that emulates the detach feature of screen.
    Copyright (C) 2001, 2004-2016 Ned T. Crigler

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#ifndef nbtty_h
#define nbtty_h

#include <config.h>

/*
** The master sends a simple stream of text to the attaching clients, without
** any protocol. This might change back to the packet based protocol in the
** future. In the meantime, however, we minimize the amount of data sent back
** and forth between the client and the master. BUFSIZE is the size of the
** buffer used for the text stream.
*/
#define BUFSIZE 4096

/* This hopefully moves to the bottom of the screen */
#define EOS "\033[999H"

int attach_main(int s, const char *ttypath, int wait_input);
int master_main(char **argv, int s);

#endif


attach.c
/*
    dtach - A simple program that emulates the detach feature of screen.
    Copyright (C) 2004-2016 Ned T. Crigler

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#include "nbtty.h"

#include <err.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/select.h>
#include <termios.h>
#include <unistd.h>

#ifndef VDISABLE
#ifdef _POSIX_VDISABLE
#define VDISABLE _POSIX_VDISABLE
#else
#define VDISABLE 0377
#endif
#endif

/*
** The original terminal settings. On exit, we restore this.
*/
static struct termios orig_term;

static int tty_in = STDIN_FILENO;
static int tty_out = STDOUT_FILENO;

static int terminal_active = 1;

/* Ignore the return code of write. This works around a compiler warning */
static ssize_t write_buffer(int fd, const unsigned char *buffer, size_t len)
{
    if (terminal_active)
        return write(fd, buffer, len);
    else
        return (ssize_t) len;
}

static ssize_t write_string(int fd, const char *str)
{
    return write_buffer(fd, (const unsigned char *) str, strlen(str));
}


/* Restores the original terminal settings. */
static void restore_term(void)
{
    tcsetattr(tty_in, TCSADRAIN, &orig_term);

    /* Make cursor visible. Assumes VT100. */
    write_string(tty_out, "\033[?25h");
}

/* Signal */
static void die(int sig)
{
    (void) sig;
    write_string(tty_out, EOS "\r\n[nbtty: terminating via signal]\r\n");
    exit(EXIT_FAILURE);
}

static void open_tty(const char *ttypath)
{
    // If already open, the close the handle.
    if (tty_in != STDIN_FILENO)
        close(tty_in);

    if (ttypath == NULL || strcmp(ttypath, "-") == 0) {
        // Check if we're using stdin
        tty_in = STDIN_FILENO;
        tty_out = STDOUT_FILENO;
    } else {
        // Open the tty or retry until it works
        for (;;) {
            int fd = open(ttypath, O_RDWR | O_CLOEXEC);
            if (fd >= 0) {
                tty_in = fd;
                tty_out = fd;
                break;
            }

            // Try again in a second?
            sleep(1);
        }
    }

    /* Save the original terminal settings. */
    if (tcgetattr(tty_in, &orig_term) < 0)
        errx(EXIT_FAILURE, "Attaching to a session requires a terminal.");

    /* Set raw mode. */
    struct termios new_term = orig_term;
    new_term.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL);
    new_term.c_iflag &= ~(IXON | IXOFF);
    new_term.c_oflag &= ~(OPOST);
    new_term.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
    new_term.c_cflag &= ~(CSIZE | PARENB);
    new_term.c_cflag |= CS8;
    new_term.c_cc[VLNEXT] = VDISABLE;
    new_term.c_cc[VMIN] = 1;
    new_term.c_cc[VTIME] = 0;
    tcsetattr(tty_in, TCSADRAIN, &new_term);
}

int attach_main(int s, const char *ttypath, int wait_input)
{
    terminal_active = !wait_input;

    /* Set some signals. */
    signal(SIGPIPE, SIG_IGN);
    signal(SIGXFSZ, SIG_IGN);
    signal(SIGHUP, die);
    signal(SIGTERM, die);
    signal(SIGINT, die);
    signal(SIGQUIT, die);

    open_tty(ttypath);

    /* Set a trap to restore the terminal when we die. */
    atexit(restore_term);

    /* Wait for things to happen */
    for (;;) {
        unsigned char buf[BUFSIZE];
        fd_set readfds;
        FD_ZERO(&readfds);
        FD_SET(tty_in, &readfds);
        FD_SET(s, &readfds);
        int highest_fd = tty_in > s ? tty_in : s;
        int rc = select(highest_fd + 1, &readfds, NULL, NULL, NULL);
        if (rc < 0) {
            if (errno != EINTR) {
                write_string(tty_out, EOS "\r\n[nbtty: select failed]\r\n");
                exit(EXIT_FAILURE);
            }
            continue;
        }

        /* Pty activity */
        if (FD_ISSET(s, &readfds)) {
            ssize_t len = read(s, buf, sizeof(buf));

            if (len == 0) {
                write_string(tty_out, EOS "\r\n[nbtty: terminating]\r\n");
                exit(EXIT_SUCCESS);
            } else if (len < 0) {
                write_string(tty_out, EOS "\r\n[nbtty: read returned an error]\r\n");
                exit(EXIT_FAILURE);
            }
            /* Send the data to the terminal. */
            write_buffer(tty_out, buf, (size_t) len);
        }

        /* User activity */
        if (FD_ISSET(tty_in, &readfds)) {
            ssize_t len = read(tty_in, buf, sizeof(buf));
            if (len <= 0) {
                if (tty_in == STDIN_FILENO)
                    exit(EXIT_FAILURE);
                open_tty(ttypath);
                continue;
            }

            if (terminal_active) {
                write_buffer(s, buf, (size_t) len);
            } else if (memchr(buf, '\r', (size_t) len)) {
                /* Activate the terminal on carriage return */
                terminal_active = 1;
                write_string(tty_out, EOS "\r\n");
            }
        }
    }
}


master.c
/*
    dtach - A simple program that emulates the detach feature of screen.
    Copyright (C) 2004-2016 Ned T. Crigler

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#include "nbtty.h"
#include "ansi.h"

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <pty.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <termios.h>
#include <time.h>
#include <unistd.h>

#include <sys/select.h>
#include <sys/wait.h>

//#define REPORT_BYTES_DROPPED

/* The pty struct - The pty information is stored here. */
struct pty {
    /* File descriptor of the pty */
    int fd;
    /* Process id of the child. */
    pid_t pid;
    /* The current window size of the pty. */
    struct winsize ws;
};

/* The connected client. */
static int client_fd = -1;

/* The pseudo-terminal created for the child process. */
static struct pty the_pty;
static uint32_t next_poll_time = 0;

/* This gets set to true when it's time to poll the window size again */
static int poll_window_size = 0;

#ifdef REPORT_BYTES_DROPPED
int bytes_dropped = 0;
#endif

static uint32_t now()
{
    static uint32_t counter = 0;
    struct timespec tp;
    if (clock_gettime(CLOCK_MONOTONIC, &tp) < 0)
        return counter++;
    return (uint32_t) tp.tv_sec;
}

/* Signal */
static void die(int sig)
{
    /* Well, the child died. */
    if (sig == SIGCHLD) {
        /* Call waitpid to avoid init having to reap a zombie orphan. (Reduces
         * noise when debugging erlinit) */
        waitpid(the_pty.pid, NULL, WNOHANG);

        /* Clean up the pty by closing it (Technically unnecessary since we're exiting shortly) */
        close(the_pty.fd);
    }

    exit(EXIT_FAILURE);
}

/* Initialize the pty structure. */
static int init_pty(char **argv)
{
    memset(&the_pty.ws, 0, sizeof(struct winsize));

    /* Create the pty process */
    the_pty.pid = forkpty(&the_pty.fd, NULL, NULL, NULL);
    if (the_pty.pid < 0)
        return -1;
    else if (the_pty.pid == 0) {
        /* Child.. Execute the program. */
        execvp(*argv, argv);

        printf(EOS "Could not execute %s: %s\r\n",
               *argv, strerror(errno));
        fflush(stdout);
        _exit(127);
    }
    /* Parent.. Finish up and return */
    return 0;
}

/* Process activity on the pty - Input and terminal changes are sent out to
** the attached clients. If the pty goes away, we die. */
static void pty_activity()
{
    unsigned char buf[BUFSIZE];
    ssize_t len;

    size_t bytes_to_read = sizeof(buf);
    if (poll_window_size)
        bytes_to_read -= ANSI_MAX_RESPONSE_LEN;

    /* Read the pty activity */
    len = read(the_pty.fd, buf, bytes_to_read);

    /* Error -> die */
    if (len <= 0)
        exit(EXIT_FAILURE);

    /* If we need to poll the window size, tack the request onto the buffer */
    if (poll_window_size) {
        len += ansi_size_request(&buf[len]);
        poll_window_size = 0;
    }

    ssize_t written = 0;
    int retries = 0;
    for (;;) {
#ifdef REPORT_BYTES_DROPPED
        {
            static int last_bytes_dropped = 0;
            if (last_bytes_dropped != bytes_dropped)
            {
                char str[64];
                sprintf(str, "[%d dropped]", bytes_dropped);
                if (write(client_fd, str, strlen(str)) > 0)
                    last_bytes_dropped = bytes_dropped;
            }
        }
#endif
        ssize_t n = write(client_fd, buf + written, len - written);
        if (n > 0) {
            written += n;
            if (written == len)
                break;

            retries++;
            if (retries > 1)
                break;
        } else if (n < 0 && errno == EINTR)
            continue;
        else
            break;
    }

#ifdef REPORT_BYTES_DROPPED
    bytes_dropped += (len - written);
#endif
}

/* Process activity from a client. */
static void client_activity()
{
    unsigned char buf[BUFSIZE];

    /* Read the activity. */
    ssize_t len = read(client_fd, buf, sizeof(buf) - ANSI_MAX_RESPONSE_LEN);
    if (len < 0 && (errno == EAGAIN || errno == EINTR))
        return;

    /* Close the client on an error. */
    if (len <= 0) {
        close(client_fd);
        client_fd = -1;
        return;
    }

    /* Check if we should poll the window size */
    if (memchr(buf, '\r', len) != NULL) {
        uint32_t current_seconds = now();

        /* poll the window size at most every 5 seconds */
        if (next_poll_time - current_seconds > 5) {
            poll_window_size = 1;
            next_poll_time = current_seconds + 5;
        }
    }

    /* Push out data to the program. */
    unsigned char output[BUFSIZE];
    size_t output_size;
    if (ansi_process_input(buf, len, output, &output_size, &the_pty.ws))
        ioctl(the_pty.fd, TIOCSWINSZ, &the_pty.ws);
    if (output_size > 0)
        write(the_pty.fd, output, output_size);
}

/* The master process - It watches over the pty process and the attached */
/* clients. */
static void master_process(char **argv)
{
    /* Okay, disassociate ourselves from the original terminal, as we
    ** don't care what happens to it. */
    setsid();

    /* Create a pty in which the process is running. */
    signal(SIGCHLD, die);
    if (init_pty(argv) < 0) {
        if (errno == ENOENT)
            errx(EXIT_FAILURE, "Could not find a pty.");
        else
            err(EXIT_FAILURE, "init_pty");
    }

    /* Set up some signals. */
    signal(SIGPIPE, SIG_IGN);
    signal(SIGXFSZ, SIG_IGN);
    signal(SIGHUP, SIG_IGN);
    signal(SIGTTIN, SIG_IGN);
    signal(SIGTTOU, SIG_IGN);
    signal(SIGINT, die);
    signal(SIGTERM, die);

    /* Make sure stdin/stdout/stderr point to /dev/null. We are now a
    ** daemon. */
    int nullfd = open("/dev/null", O_RDWR);
    dup2(nullfd, 0);
    dup2(nullfd, 1);
    dup2(nullfd, 2);
    if (nullfd > 2)
        close(nullfd);

    /* Loop forever. */
    while (1) {
        /* Re-initialize the file descriptor set for select. */
        fd_set readfds;
        FD_ZERO(&readfds);
        FD_SET(client_fd, &readfds);
        FD_SET(the_pty.fd, &readfds);
        int highest_fd = client_fd > the_pty.fd ? client_fd : the_pty.fd;

        /* Wait for something to happen. */
        if (select(highest_fd + 1, &readfds, NULL, NULL, NULL) < 0) {
            if (errno == EINTR)
                continue;
            exit(EXIT_FAILURE);
        }

        /* Activity on a client? */
        if (FD_ISSET(client_fd, &readfds))
            client_activity();
        /* pty activity? */
        if (FD_ISSET(the_pty.fd, &readfds))
            pty_activity();
    }
}

int master_main(char **argv, int s)
{
    ansi_reset_parser();

    if (fcntl(s, F_SETFD, FD_CLOEXEC) < 0)
        err(EXIT_FAILURE, "fcnt(F_SETFD, FD_CLOEXEC)");

    int flags = fcntl(s, F_GETFL);
    if (flags < 0 || fcntl(s, F_SETFL, flags | O_NONBLOCK) < 0)
        err(EXIT_FAILURE, "fcnt(F_SETFL, 0x%x | O_NONBLOCK)", flags);

    client_fd = s;

    /* Fork off so we can daemonize and such */
    pid_t pid = fork();
    if (pid < 0) {
        err(EXIT_FAILURE, "fork");
    } else if (pid == 0) {
        /* Child - this becomes the master */
        master_process(argv);
        return 0;
    }
    /* Parent - just return. */
    return 0;
}


main.c
/*
    nbtty - A subset of dtach that makes the terminal nonblocking
    Copyright (C) 2004-2016 Ned T. Crigler

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
#include "nbtty.h"

#include <err.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <syslog.h>
#include <termios.h>
#include <unistd.h>

/* Make sure the binary has a copyright. */
const char copyright[] = "nbtty - version 0.3.0 (C)Copyright 2004-2016 Ned T. Crigler, 2017 Frank Hunleth";

static void usage()
{
    errx(EXIT_FAILURE, "nbtty [--tty <path>|--wait-input] <command> [args...]");
}

int main(int argc, char **argv)
{
    const char *ttypath = NULL;
    int wait_input = 0;

    for (;;) {
        static struct option long_options[] = {
            {"tty",     required_argument, 0,  't' },
            {"wait-input",  no_argument,   0,  'w' },
            {0,         0,                 0,  0 }
        };

        int c = getopt_long(argc, argv, "+tw", long_options, NULL);
        if (c == -1)
            break;

        switch (c) {
        case 't':
            ttypath = optarg;
            break;

        case 'w':
            wait_input = 1;
            break;
        default:
            usage();
        }
     }

    if (optind == argc)
        usage();

    int sv[2];
    if (socketpair(AF_UNIX, SOCK_STREAM, 0, sv) < 0)
        err(EXIT_FAILURE, "socketpair");

    if (master_main(&argv[optind], sv[0]) != 0)
        return 1;

    close(sv[0]);

    return attach_main(sv[1], ttypath, wait_input);
}
